### ISSUE TITLE
"Flat" install for local trees

### ISSUE THREAD

**Issue by mcclure** [luarocks/luarocks#835]
When installing or making with --tree, it should be possible to install in a "flat" fashion where instead of all the `share/lua/5.1`/`app/lib/luarocks/rocks-5.1/luigi/local-1/luigi` stuff the files are just copied directly into the target directory. "Collisions" (IE a single command attempting to install two separate versions/namespaces of the same package at once) should be detected and result in a warning or error. Metadata such as manifests can possibly be omitted entirely in this case.

The use case for this is platforms such as Löve, where an entire program and all its dependencies are distributed as a single directory. Currently the user must do manipulation of the package path to load libraries out of luarocks from Löve, a significant ease-of-use barrier given Löve is designed to be usable by beginning programmers. The benefits normally offered by the tree structure do not help in this case since the rock tree will only be accessed by a single program.

I would be willing to write code toward this if I could get some guidance on where the changes would need to be.

**Comment by hishamhm** [comment]
One first step towards this would be to try to produce the desired behavior using configuration only.

Using LuaRocks 3.0.0beta1 (or current `master`), you should be able to run `luarocks config` with no arguments and see the entire active configuration. You can override any of those entries in the `.luarocks/config-5.x.lua` file of the current project dir (initialized with `luarocks init`). What would a flat Löve-friendly config look like?

If that works out, one possible approach for this idea would be to have a "template" argument to `luarocks init` that creates a customized configuration (and even extra boilerplate files, etc) — how does that sound?

**Comment by mcclure** [comment]
If the goal is to help "non-advanced" programmers, being able to issue a specific argument with "luarocks init" or copy an existing luarocks would I think, would be superior to having to enter a complex series of --arguments on every build. I will test that.

**Comment by hishamhm** [comment]
Exactly. My suggestion is essentially to begin by prototyping a config file that looks like the one we'd like to have generated by something like `luarocks init --template=love2d`

**Comment by bjornbytes** [comment]
I was able to get a flat structure by adding `lua_modules_path = ""` to the project config file (not sure if there's a better way to do it).  The rockspecs are still nested under `lib` though.

**Comment by mcclure** [comment]
Okay so here's my little demo repo https://bitbucket.org/runhello/loopfriend

I tried bjorn's lua_modules_path = ""

There were only two problems:

- I had to put lua_modules_path in the LUAROCKS_CONFIG file. If I put it in the rockspec, I get: `Error: /Users/mcc/work/h/loopfriend/rockspec: Unknown field lua_modules_path (using rockspec format 3.0)`. Is there a way to set config values in the rockspec? Honestly my other config value (rocks_servers) would also be *more* appropriate in the rockspec.

- So this one's interesting -- One of my dependencies (I have my own little private repo, in the rocks/ dir, so I created this rockspec myself) has `copy_directories = {"luigi/theme/light", "luigi/theme/dark"}` in its rockspec. This gets copied into the `lib` directory bjorn points out! I don't really care if there's a lib directory full of rockspecs (maybe it's even handy to have what files were installed cached somewhere?)  but *resources* getting copied into lib/ doesn't work because the lua library will expect to find them at the "actual" paths (luigi/theme/light etc). Is there a way to flatten that lib dir also? `lib_modules_path` doesn't seem to do it.

**Comment by bjornbytes** [comment]
I created a demo repo at https://github.com/bjornbytes/luarocks-test with some config/script changes that work well for this sort of "portable" distribution model, including the steps taken to get there and a list of improvements that would make it easier to use.  Do any of these improvements make sense to add?  If so, I'd be interested in helping out with the implementation.
